sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/predicates.scala

```diff
@@ -335,6 +343,41 @@ case class In(value: Expression, list: Seq[Expression]) extends Predicate {
        """.stripMargin)
   }
 
+  private def genCodeWithSwitch(ctx: CodegenContext, ev: ExprCode): ExprCode = {
+    val (nullLiterals, nonNullLiterals) = list.partition {
+      case Literal(null, _) => true
+      case _ => false
+    }
+    val listGen = nonNullLiterals.map(_.genCode(ctx))
+    val valueGen = value.genCode(ctx)
+
+    val caseBranches = listGen.map(literal =>
+      s"""
+         |case ${literal.value}:
+         |  ${ev.value} = true;
+         |  break;
+       """.stripMargin)
+
+    ev.copy(code =
+      code"""
+         |${valueGen.code}
+         |${CodeGenerator.JAVA_BOOLEAN} ${ev.isNull} = ${valueGen.isNull};
+         |${CodeGenerator.JAVA_BOOLEAN} ${ev.value} = false;
+         |if (!${valueGen.isNull}) {
+         |  switch (${valueGen.value}) {
+         |    ${caseBranches.mkString("")}
+         |    default:
+         |      ${ev.isNull} = ${nullLiterals.nonEmpty};
+         |  }
+         |}
+       """.stripMargin)
+  }
+
+  private def isSwitchCompatible: Boolean = list.forall {
```

Comment:
Could you please take care of the following limitation of Java switch statement, too?
> npairs pairs of signed 32-bit values

https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-6.html#jvms-6.5.lookupswitch
https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-6.html#jvms-6.5.tableswitch

Reply:
Could you, please, elaborate a bit on this? I am not sure I got. Shouldn't we be fine if we limit this approach to bytes/shorts/ints?

Reply:
Sorry for missing some words. My comment is that `isSwitchCompatible` can be true only if `list.size` is INT.MAX. Otherwise, Janino will cause a failure.